pkg/controller/disruption/disruption.go

```diff
@@ -0,0 +1,587 @@
+	if !exists {
+		return err
+	}
+	if err != nil {
+		glog.Errorf("Unable to retrieve PodDisruptionBudget %v from store: %v", key, err)
+		return err
+	}
+
+	pdb := obj.(*policy.PodDisruptionBudget)
+
+	if err := dc.trySync(pdb); err != nil {
+		return dc.failSafe(pdb)
+	}
+
+	return nil
+}
+
+func (dc *DisruptionController) trySync(pdb *policy.PodDisruptionBudget) error {
+	pods, err := dc.getPodsForPdb(pdb)
+	if err != nil {
+		return err
+	}
+
+	expectedCount, desiredHealthy, err := dc.getExpectedPodCount(pdb, pods)
+	if err != nil {
+		return err
+	}
+
+	currentHealthy := dc.countHealthyPods(pdb, pods)
+	err = dc.updatePdbSpec(pdb, currentHealthy, desiredHealthy, expectedCount)
+
+	return err
+}
+
+func (dc *DisruptionController) getExpectedPodCount(pdb *policy.PodDisruptionBudget, pods []*api.Pod) (expectedCount, desiredHealthy int32, err error) {
+	err = nil
+	if pdb.Spec.MinAvailable.Type == intstr.Int {
+		desiredHealthy = pdb.Spec.MinAvailable.IntVal
+		expectedCount = int32(len(pods))
+	} else if pdb.Spec.MinAvailable.Type == intstr.String {
+		// When the user specifies a fraction of pods that must be available, we
+		// use as the fraction's denominator
+		// SUM_{all c in C} scale(c)
+		// where C is the union of C_p1, C_p2, ..., C_pN
+		// and each C_pi is the set of controllers controlling the pod pi
+
+		// k8s only defines what will happens when 0 or 1 controllers control a
+		// given pod.  We explicitly exclude the 0 controllers case here, and we
+		// report an error if we find a pod with more than 1 controller.  Thus in
+		// practice each C_pi is a set of exactly 1 controller.
+
+		// A mapping from controllers to their scale.
+		controllerScale := map[types.UID]int32{}
+
+		// 1. Find the controller(s) for each pod.  If any pod has 0 controllers,
+		// that's an error.  If any pod has more than 1 controller, that's also an
+		// error.
+		for _, pod := range pods {
+			controllerCount := 0
+			for _, finder := range dc.finders() {
+				var controllers []controllerAndScale
+				controllers, err = finder(pod)
+				if err != nil {
+					return
+				}
+				for _, controller := range controllers {
+					controllerScale[controller.UID] = controller.scale
+					controllerCount++
+				}
+			}
+			if controllerCount == 0 {
+				err = fmt.Errorf("Asked for percentage, but found no controllers for pod %q", pod.Name)
+				return
+			} else if controllerCount > 1 {
+				err = fmt.Errorf("Pod %q has %v>1 controllers", pod.Name, controllerCount)
+				return
+			}
+		}
+
+		// 2. Add up all the controllers.
+		expectedCount = 0
+		for _, count := range controllerScale {
+			expectedCount += count
+		}
+
+		// 3. Do the math.
+		var dh int
+		dh, err = intstr.GetValueFromIntOrPercent(&pdb.Spec.MinAvailable, int(expectedCount), true)
+		if err != nil {
+			return
+		}
+		desiredHealthy = int32(dh)
+	}
+
+	return
+}
+
+func (dc *DisruptionController) countHealthyPods(pdb *policy.PodDisruptionBudget, pods []*api.Pod) (currentHealthy int32) {
```

Comment:
No. Named returns are *good*-- self documenting.

Some people find naked 'returns' in that case confusing, so that's the thing we'd prohibit.

I personally don't find them confusing at all and don't understand why others do, but I'll go along with the will of the majority here.